system/security/plugins/jwtSecurity/README.md

```diff
@@ -0,0 +1,164 @@
+modify a user's authorization while the user is logged in.  This plugin fully supports refresh tokens by validating
+token lifetime at every authorization check and calling a ```JWT refresh service``` (also known as a JWT refresh
+endpoint) when needed.  This largely follows the OIC specification.
+
+### Deviations From OIC Specification
+
+* Initial authentication:  As stated above, the ```esp``` process will gather the username/password credentials instead
+of a third party, then send those credentials off to another service.  In a true OIC configuration, the client process
+(the ```esp``` process) never sees user credentials and relies on an external service to gather them from the user.
+* The request made to the ```JWT login service``` is a POST HTTP call containing four items in JSON format; example:
+
+```
+	{
+		"username": "my_username",
+		"password": "my_password",
+		"client_id": "https://myhpcccluster.com",
+		"nonce": "hf674DTRMd4Z1s"
+	}
+```
+* The ```JWT login service``` should reply with an OIC-compatible JSON-formatted reply.  See
+<https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse> for an example of a successful authentication and
+<https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse> for an example of an error response.  The
+token itself is comprised of several attributes, followed by HPCC Systems-specific claims; see
+<https://openid.net/specs/openid-connect-core-1_0.html#IDToken> (all required fields are indeed required, plus the nonce
+field).
+	* Note that for success replies, the ```access_token``` and ```expires_in``` values are ignored by this plugin.
+* If the token expires, the plugin will call the ```JWT refresh service``` to request a new token.  This follows
+<https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens> except that the ```client_secret``` and ```scope```
+fields in the request are omitted.
+
+### Implications of Deviations
+
+The most obvious outcome of this implementation is that a custom service/endpoint needs to be available.  Or rather two
+services:  One to handle the initial user login and one to handle token refreshes.  Neither service *precisely* handles
+requests and replies in an OIC-compatible way, but the tokens themselves *are* OIC-compatible, which is good.  That
+allows you to use third-party JWT libraries to construct and validate those tokens.
+
+### HPCC Systems Configuration Notes
+
+Several items must be defined in the platform's configuration.  Within configmgr, the ```jwtsecmgr``` Security Manager
+plugin must be added as a component and then modified according to your environment:
+
+* The URL or unique name of this HPCC Systems cluster, used as the ```client_id``` in token requests
+* Full URL to the ```JWT Login Endpoint``` (should be https, but not required)
+* Full URL to the ```JWT Refresh Endpoint``` (should be https, but not required)
```

Comment:
Should we require HTTPS for these?

Reply:
I don't think we should require https, but I could be convinced otherwise.  A sealed network environment, or a dev environment, may not want the additional complexity.  Let me know if you want this requirement added.

Reply:
If we send password hashes it may be less critical.

Reply:
I will definitely change to sending a hashed password, as noted elsewhere (I think).